File: leetcodeQuestion/leetcode53.cpp
/* 给定一个整数数组 nums ，找到一个具有最大和的连续子数组（子数组最少包含一个元素），返回其最大和。

 

示例 1：

输入：nums = [-2,1,-3,4,-1,2,1,-5,4]
输出：6
解释：连续子数组 [4,-1,2,1] 的和最大，为 6 。 */

#include<iostream>
#include<vector>
#include<math.h>
using namespace std;

class Solution
{
public:
    int maxSubArray(vector<int>& nums) {
      vector<int> subArray(nums.size(), 0);
      int maxSum = nums[0];
      subArray[0] = nums[0];
      for (int i=1;i<nums.size();i++){
        subArray[i] = max(subArray[i - 1] + nums[i], nums[i]);
        maxSum = max(maxSum, subArray[i]);
      }
      return maxSum;

    }
};

File: leetcodeQuestion/leetcode/src/leetcode53/Leetcode53.java
public class Leetcode53 {
  public int maxSubArray(int[] nums) {
    int _l = nums.length;
    int[] dp = new int[_l];
    dp[0] = nums[0];
    int maxSum = dp[0];
    for (int i = 1; i < _l; i++) {
      dp[i] = Math.max(nums[i], nums[i] + dp[i - 1]);
      maxSum = Math.max(maxSum, dp[i]);
    }

    return maxSum;
  }

  public static void main(String[] args) {
    Leetcode53 demo = new Leetcode53();
    int[] nums = { -2, 1, -3, 4, -1, 2, 1, -5, 4 };
    System.out.println(demo.maxSubArray(nums));
  }
}
